File: CS/DevExpressWalkthrough/ViewModels/Department/DepartmentViewModel.cs
using System;
using System.Linq;
using System.Linq.Expressions;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using DevExpress.Mvvm;
using DevExpress.Mvvm.POCO;
using DevExpressWalkthrough.Common.Utils;
using DevExpressWalkthrough.DepartmentContextDataModel;
using DevExpressWalkthrough.Common.DataModel;
using DevExpressWalkthrough;
using DevExpressWalkthrough.Common.ViewModel;

namespace DevExpressWalkthrough.ViewModels {
    /// <summary>
    /// Represents the single Department object view model.
    /// </summary>
    public partial class DepartmentViewModel : SingleObjectViewModel<Department, int, IDepartmentContextUnitOfWork> {

        /// <summary>
        /// Creates a new instance of DepartmentViewModel as a POCO view model.
        /// </summary>
        /// <param name="unitOfWorkFactory">A factory used to create a unit of work instance.</param>
        public static DepartmentViewModel Create(IUnitOfWorkFactory<IDepartmentContextUnitOfWork> unitOfWorkFactory = null) {
            return ViewModelSource.Create(() => new DepartmentViewModel(unitOfWorkFactory));
        }

        /// <summary>
        /// Initializes a new instance of the DepartmentViewModel class.
        /// This constructor is declared protected to avoid undesired instantiation of the DepartmentViewModel type without the POCO proxy factory.
        /// </summary>
        /// <param name="unitOfWorkFactory">A factory used to create a unit of work instance.</param>
        protected DepartmentViewModel(IUnitOfWorkFactory<IDepartmentContextUnitOfWork> unitOfWorkFactory = null)
            : base(unitOfWorkFactory ?? UnitOfWorkSource.GetUnitOfWorkFactory(), x => x.Departments, x => x.Name) {
        }

        /// <summary>
        /// The view model for the DepartmentCourses detail collection.
        /// </summary>
        public CollectionViewModel<Course, int, IDepartmentContextUnitOfWork> DepartmentCoursesDetails
        {
            get { return GetDetailsCollectionViewModel((DepartmentViewModel x) => x.DepartmentCoursesDetails, x => x.Courses, x => x.DepartmentID, (x, key) => { x.DepartmentID = key; }); }
        }

        /// <summary>
        /// The view model for the DepartmentEmployees detail collection.
        /// </summary>
        public CollectionViewModel<Employee, int, IDepartmentContextUnitOfWork> DepartmentEmployeesDetails
        {
            get { return GetDetailsCollectionViewModel((DepartmentViewModel x) => x.DepartmentEmployeesDetails, x => x.Employees, x => x.DepartmentID, (x, key) => { x.DepartmentID = key; }); }
        }
    }
}

File: VB/DevExpressWalkthrough.VB/ViewModels/Department/DepartmentViewModel.vb
Imports System
Imports System.Linq
Imports System.Linq.Expressions
Imports System.Collections.Generic
Imports System.Collections.ObjectModel
Imports DevExpress.Mvvm
Imports DevExpress.Mvvm.POCO
Imports DevExpressWalkthrough.VB.Common.Utils
Imports DevExpressWalkthrough.VB.DepartmentContextDataModel
Imports DevExpressWalkthrough.VB.Common.DataModel
Imports DevExpressWalkthrough.VB
Imports DevExpressWalkthrough.VB.Common.ViewModel
Namespace Global.DevExpressWalkthrough.VB.ViewModels
    ''' <summary>
    ''' Represents the single Department object view model.
    ''' </summary>
    Public Partial Class DepartmentViewModel
        Inherits SingleObjectViewModel(Of Department, Integer, IDepartmentContextUnitOfWork)
        ''' <summary>
        ''' Creates a new instance of DepartmentViewModel as a POCO view model.
        ''' </summary>
        ''' <param name="unitOfWorkFactory">A factory used to create a unit of work instance.</param>
        Public Shared Function Create(Optional ByVal unitOfWorkFactory As IUnitOfWorkFactory(Of IDepartmentContextUnitOfWork) = Nothing) As DepartmentViewModel
            Return ViewModelSource.Create(Function() New DepartmentViewModel(unitOfWorkFactory))
        End Function
        ''' <summary>
        ''' Initializes a new instance of the DepartmentViewModel class.
        ''' This constructor is declared protected to avoid undesired instantiation of the DepartmentViewModel type without the POCO proxy factory.
        ''' </summary>
        ''' <param name="unitOfWorkFactory">A factory used to create a unit of work instance.</param>
        Protected Sub New(Optional ByVal unitOfWorkFactory As IUnitOfWorkFactory(Of IDepartmentContextUnitOfWork) = Nothing)
            MyBase.New(If(unitOfWorkFactory, UnitOfWorkSource.GetUnitOfWorkFactory()), Function(ByVal x) x.Departments, Function(ByVal x) x.Name)
        End Sub
        ''' <summary>
        ''' The view model for the DepartmentCourses detail collection.
        ''' </summary>
        Public ReadOnly Property DepartmentCoursesDetails As CollectionViewModel(Of Course, Integer, IDepartmentContextUnitOfWork)
            Get
                Return GetDetailsCollectionViewModel(Function(ByVal x As DepartmentViewModel) x.DepartmentCoursesDetails, Function(ByVal x) x.Courses, Function(ByVal x) x.DepartmentID, Sub(ByVal x, ByVal key) x.DepartmentID = key)
            End Get
        End Property
        ''' <summary>
        ''' The view model for the DepartmentEmployees detail collection.
        ''' </summary>
        Public ReadOnly Property DepartmentEmployeesDetails As CollectionViewModel(Of Employee, Integer, IDepartmentContextUnitOfWork)
            Get
                Return GetDetailsCollectionViewModel(Function(ByVal x As DepartmentViewModel) x.DepartmentEmployeesDetails, Function(ByVal x) x.Employees, Function(ByVal x) x.DepartmentID, Sub(ByVal x, ByVal key) x.DepartmentID = key)
            End Get
        End Property
    End Class
End Namespace
